HACKING-URL.md

```diff
@@ -0,0 +1,319 @@
+and at the commit messages in the pull requests. Do the same.
+If any doubt, get back to this guide which should contain the answer.
+
+Our URL handling has two main parts: editing handlers (unwanted URL block, URL
+cleanup, relationship type autoselection, relationship validation), and
+display handlers (with favicons, in relationship lists, in the sidebar).
+
+Table of Contents
+-----------------
+
+<!-- toc -->
+
+- [Prerequisites](#prerequisites)
+- [URL editing handlers](#url-editing-handlers)
+  * [Cleanup](#cleanup)
+  * [Autoselection](#autoselection)
+  * [Validation](#validation)
+  * [Supporting new relationship types](#supporting-new-relationship-types)
+  * [Entity-wide rules](#entity-wide-rules)
+- [URL display handlers](#url-display-handlers)
+  * [Favicons](#favicons)
+  * [Sidebar display](#sidebar-display)
+  * [Rendered scheme override](#rendered-scheme-override)
+  * [Custom in-page display](#custom-in-page-display)
+  * [Miscellaneous display customization](miscellaneous-display-customization)
+
+<!-- tocstop -->
+
+Prerequisites
+-------------
+
+Allowing to link external websites as
+either [other databases](https://musicbrainz.org/doc/Other_Databases_Relationship_Type/Whitelist) (for persistence issues)
+or [lyrics websites](https://musicbrainz.org/doc/Style/Relationships/URLs/Lyrics_whitelist) (for license issues)
+must be approved by a STYLE decision beforehand.
+See <https://musicbrainz.org/doc/Proposals>
+for how to request such STYLE decision,
+before even thinking about coding handlers.
+
+Other relationship types don’t require any STYLE decision,
+but having a self-assigned MBS ticket is still required.
+
+The following research work is usually expected about external websites to be handled:
+
+* Form and number of its URLs in the MusicBrainz database already;
+  - Tip: For a first approximate count, you can query the MusicBrainz API about URLs with (approximate) patterns;
+  - For example [`url:*/openlibrary.org/*`](https://musicbrainz.org/ws/2/url?query=url:*/openlibrary.org/*&limit=1&fmt=json) counts Open Library URLs (mainly).
+
+* Mapping between patterns of its URLs and MusicBrainz entities with relationship type(s);
+  - For example `/authors/` Open Library URLs match MusicBrainz artists.
+
+* Potential alternative forms of its URLs;
+  - For example `youtu.be/REF` is an alias for `www.youtube.com/watch?v=REF`.
+
+* Potential query (`?...`) and fragment (`#...`);
+  - For example `&list=...` is removed from YouTube video URLs taken from a playlist.
+
+* Potential patterns to be blocked (search pages, unreliable URLs...);
+  - For example `youtube.com/playlist` URLs are unwanted and blocked with a message.
+
+* Rationale and real examples for all the above.
+
+Expected behavior and rationale should be given in the ticket,
+and implementation details in the pull request and commit messages.
+
+It isn’t required to make a fine-grained implementation.
+Sometimes it is even impossible because of URL patterns.
```

Comment:
I have been intentionally vague about what  “a fine-grained implementation” is. It is way more general than just the validation. It is about everything in this document.

Reply:
Then maybe just:

```suggestion
You don't always have to write a fine-grained implementation that looks at every single detail.
In fact, sometimes it is not even possible to do so, for example because the URL patterns
are not distinctive and do not allow for very precise validation.
```

I want to make sure it's not read as "you can just do some lazy thing and not even look at half this stuff" though :D

Reply:
> "you can just do some lazy thing and not even look at half this stuff"

This is pretty much what I meant. There is no need to make a full-fledged implementation for every website. It mostly depends on the level of involvement of the contributor. It has always been. 

That reminds me that there is at least one exception to this in the case where the site is linked using a restricted relationship type; Then it requires a full-fledged implementation otherwise the restriction would not make sense any longer.

Reply:
The “exception” is “the case where the site is linked using a restricted relationship type” as in [`RESTRICTED_LINK_TYPES`](https://github.com/metabrainz/musicbrainz-server/blob/v-2023-01-10-hotfixes/root/static/scripts/edit/URLCleanup.js#L300-L327); which indeed “requires a full-fledged implementation” (relationship type autoselection, URL cleanup/normalization, relationship validation, sidebar display).

I’m fine with removing this paragraph altogether as it doesn’t add much.

Reply:
Fine with just removing it.